tools/print_tock_memory_usage.py

```diff
@@ -0,0 +1,432 @@
+            if symbol[0:6] == '.Lanon' or symbol[0:5] == "anon." or symbol[0:4] == 'str.':
+                key = "Constant strings"
+            elif symbol[0:8] == ".hidden ":
+                key = "ARM aeabi support"
+            elif symbol[0:3] == "_ZN":
+                key = "Unidentified auto-generated"
+            else:
+                key = "Unmangled globals (C-like code)"
+                name = symbol
+        else:
+            # Packages have a trailing :: while other categories don't;
+            # this allows us to disambiguate when * is relevant or not
+            # in printing.
+            key = "::".join(tokens[0:symbol_depth]) + "::"
+            name = "::".join(tokens[symbol_depth:])
+
+            if key in groups.keys():
+                groups[key].append((name, size))
+            else:
+                groups[key] = [(name, size)]
+
+        # Set state for next iteration
+        expected_addr = addr + size
+        prev_symbol = symbol
+
+    if waste and waste_sum > 0:
+        output = output + "Total of " + str(waste_sum) + " bytes wasted in " + section + "\n"
+        
+    return output
+        
+def string_for_group(key, padding_size, group_size, num_elements):
+    """Return the string for a group of variables, with padding added on the
+       right; decides whether to add a * or not based on the name of the group
+       and number of elements in it."""
+    if num_elements == 1: # If there's a single symbol (a variable), print it.
+        key = key[:-2]
+        key = key + ":"
+        key = key.ljust(padding_size + 2, ' ')
+        return ("  " + key + str(group_size) + " bytes\n")
+    else: # If there's more than one, print the key as a namespace
+        if key[-2:] == "::":
+            key = key + "*"
+            key = key.ljust(padding_size + 2, ' ')
+            return ("  " + key + str(group_size) + " bytes\n")
+        else:
+            key = key + ":"
+            key = key.ljust(padding_size + 2, ' ')
+            return ("  " + key + str(group_size) + " bytes\n")
+
+def print_groups(title, groups):
+    """Print title, then all of the variable groups in groups."""
+    group_sum = 0
+    output = ""
+    max_string_len = len(max(groups.keys(), key=len))
+    for key in sorted(groups.keys()):
+        symbols = groups[key]
+
+        group_size = 0
+        for (_, size) in symbols:
+            group_size = group_size + size
+
+        output = output + string_for_group(key, max_string_len, group_size, len(symbols))
+        group_sum = group_sum + group_size
+
+    print(title + ": " + str(group_sum) + " bytes")
+    print(output, end = '')
+
+def print_symbol_information():
+    """Print out all of the variable and function groups with their flash/RAM
+       use."""
+    variable_groups = {}
+    gaps = group_symbols(variable_groups, kernel_initialized, show_waste, "RAM")
+    gaps = gaps + group_symbols(variable_groups, kernel_uninitialized, show_waste, "Flash+RAM")
```

Comment:
Should this be "Flash+RAM" or just "RAM"?  It is printed under the "Variable groups (RAM)" section and does not seem to include the bytes of wasted flash from functions groups.